drv/stm32h7-update-server/src/main.rs

```diff
@@ -0,0 +1,342 @@
+        }
+
+        let start = BANK_ADDR + (word_number * FLASH_WORD_BYTES) as u32;
+
+        if bytes.len() != FLASH_WORD_BYTES {
+            return Err(UpdateError::BadLength.into());
+        }
+
+        for i in 0..FLASH_CPU_WORDS {
+            unsafe {
+                let byte_offset = i * 4;
+                let mut word: [u8; 4] = [0; 4];
+                word.clone_from_slice(&bytes[byte_offset..byte_offset + 4]);
+
+                core::ptr::write_volatile(
+                    (start + (i * 4) as u32) as *mut u32,
+                    u32::from_ne_bytes(word),
+                );
+            }
+        }
+
+        let b = self.poll_flash_done();
+        ringbuf_entry!(Trace::WriteEnd);
+        b
+    }
+
+    // All sequences can be found in RM0433 Rev 7
+    fn unlock(&mut self) {
+        if !self.flash.bank2().cr.read().lock().bit() {
+            return;
+        }
+
+        self.flash
+            .bank2()
+            .keyr
+            .write(|w| unsafe { w.keyr().bits(FLASH_KEY1) });
+        self.flash
+            .bank2()
+            .keyr
+            .write(|w| unsafe { w.keyr().bits(FLASH_KEY2) });
+
+        self.flash
+            .optkeyr()
+            .write(|w| unsafe { w.optkeyr().bits(FLASH_OPT_KEY1) });
+        self.flash
+            .optkeyr()
+            .write(|w| unsafe { w.optkeyr().bits(FLASH_OPT_KEY2) });
+    }
+
+    fn bank_erase(&mut self) -> Result<(), RequestError<UpdateError>> {
+        ringbuf_entry!(Trace::EraseStart);
+        self.flash
+            .bank2()
+            .cr
+            .write(|w| w.start().set_bit().ber().set_bit());
+
+        let b = self.poll_flash_done();
+        ringbuf_entry!(Trace::EraseEnd);
+        b
+    }
+}
+
+impl idl::InOrderUpdateImpl for ServerImpl<'_> {
+    fn prep_image_update(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError<UpdateError>> {
+        match self.state {
+            UpdateState::InProgress | UpdateState::Finished => {
+                return Err(UpdateError::UpdateInProgress.into())
+            }
+            _ => (),
+        }
+
+        self.unlock();
+        self.bank_erase()?;
+        self.state = UpdateState::InProgress;
+        Ok(())
+    }
+
+    fn write_one_block(
+        &mut self,
+        _: &RecvMessage,
+        block_num: usize,
+        block: LenLimit<Leased<R, [u8]>, BLOCK_SIZE_BYTES>,
+    ) -> Result<(), RequestError<UpdateError>> {
+        match self.state {
+            UpdateState::NoUpdate | UpdateState::Finished => {
+                return Err(UpdateError::UpdateInProgress.into())
+            }
+            _ => (),
+        }
+
+        let len = block.len();
+        let mut flash_page: [u8; BLOCK_SIZE_BYTES] = [0; BLOCK_SIZE_BYTES];
+
+        block
+            .read_range(0..len as usize, &mut flash_page)
+            .map_err(|_| RequestError::Fail(ClientError::WentAway))?;
+
+        if len != BLOCK_SIZE_BYTES {
```

Comment:
Okay, I _think_ I see what this is doing, but I think it could use a comment. I suggest

- Changing this check to `<` to make it clear that you're filling in the tail of any too-short block.
- Add a comment stating that.
- Change the while-loop-acting-as-a-for-loop to:

```rust
flash_page[len..].fill(0);
```

With that alternative implementation you can actually lose the check on line 279 if you want.